ferrocene/tools/self-test/src/linkers/argparse.rs

```diff
@@ -0,0 +1,225 @@
+            output.push(result);
+        } else if let Some(result) = long_opt("EL", arg, || LinkerArg::LittleEndian) {
+            output.push(result);
+        } else if let Some(result) = long_opt("pie", arg, || LinkerArg::PicExecutable) {
+            output.push(result);
+        } else if let Some(result) = long_opt("pic-executable", arg, || LinkerArg::PicExecutable) {
+            output.push(result);
+        } else if let Some(result) = long_opt("no-pie", arg, || LinkerArg::NonPicExecutable) {
+            output.push(result);
+        } else if let Some(result) = short_arg('I', arg, &mut args_iter, LinkerArg::DynamicLinker) {
+            output.push(result);
+        } else if let Some(result) = long_arg("dynamic-linker", arg, &mut args_iter, LinkerArg::DynamicLinker) {
+            output.push(result);
+        } else if let Some(result) = long_arg("sysroot", arg, &mut args_iter, LinkerArg::Sysroot) {
+            output.push(result);
+        } else if let Some(result) = long_opt("build-id", arg, || LinkerArg::BuildId) {
+            output.push(result);
+        } else if let Some(result) = long_opt("eh-frame-hdr", arg, || LinkerArg::EhFrameHeader) {
+            output.push(result);
+        } else if let Some(result) = long_arg("hash-style", arg, &mut args_iter, LinkerArg::HashStyle) {
+            output.push(result);
+        } else if let Some(result) = long_opt("as-needed", arg, || LinkerArg::AsNeeded) {
+            output.push(result);
+        } else if let Some(result) = long_opt("no-as-needed", arg, || LinkerArg::NoAsNeeded) {
+            output.push(result);
+        } else if let Some(result) = long_opt("push-state", arg, || LinkerArg::PushState) {
+            output.push(result);
+        } else if let Some(result) = long_opt("pop-state", arg, || LinkerArg::PopState) {
+            output.push(result);
+        } else if let Some(result) = long_opt("fix-cortex-a53-843419", arg, || LinkerArg::FixCortexA53_843419) {
+            output.push(result);
+        } else if arg.starts_with("-") {
+            output.push(LinkerArg::Unknown(arg));
+        } else {
+            output.push(LinkerArg::Input(arg));
+        }
+    }
+    output
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_rationalise_linker_args() {
+        // these are the ones we saw from gcc
+        let input = include_str!("snapshots/ferrocene_self_test__linkers__argparse__tests__rationalise_linker_args.input");
+        let output = rationalise_linker_args(input.lines());
+        insta::assert_debug_snapshot!(output);
+    }
+
+    #[test]
+    fn test_rationalise_linker_extra_args() {
+        // these are extra ones we added in case we see them
+        let input = include_str!("snapshots/ferrocene_self_test__linkers__argparse__tests__rationalise_linker_extra_args.input");
+        let output = rationalise_linker_args(input.lines());
+        insta::assert_debug_snapshot!(output);
+    }
```

Comment:
Might be worth configuring insta [to avoid prepending the module name to the test files](https://docs.rs/insta/latest/insta/struct.Settings.html#method.set_prepend_module_to_snapshot), to avoid huge file names. This should do (around each assert):

```rust
insta::with_settings!({ prepend_module_to_snapshot: false }, {
    insta::assert...
});
```

Reply:
It's `{foo => value}`, but yes, done :)